heron/common/src/java/com/twitter/heron/common/basics/WakeableLooper.java

```diff
@@ -44,36 +44,44 @@
   private final PriorityQueue<TimerTask> timers;
 
   // The tasks would be invoked before exit
-  private final ArrayList<Runnable> exitTasks;
+  private final List<Runnable> exitTasks;
 
   // For selector since there is bug in selector.select(timeout): we could not
   // use a timeout > 10 * Integer.MAX_VALUE
   // So here we set Integer.MAX_VALUE as the infinite future
   // We will also multiple 1000*1000 to convert mill-seconds to nano-seconds
   private static final Duration INFINITE_FUTURE = Duration.ofMillis(Integer.MAX_VALUE);
   private volatile boolean exitLoop;
+  private boolean clearTasksOnWakeup = false;
```

Comment:
maybe better for the two flags to be initialized in constructor?

Another suggest is to rename to clearTasksOnWakeupFlag to avoid variable and function sharing the same name. Maybe not a big deal. Your call.

Reply:
I have renamed